src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2391,12 +2393,20 @@ public ReflectionPatternContext (LinkContext context, MethodDefinition methodCal
 				MethodCalling = methodCalling;
 				MethodCalled = methodCalled;
 
+				_patternAnalysisAttempted = false;
 				_patternReported = false;
 			}
 
+			public void AnalyzingPattern ()
+			{
+				_patternAnalysisAttempted = true;
+			}
+
 			public void RecordRecognizedPattern<T> (T accessedItem, Action mark)
 				where T : IMemberDefinition
 			{
+				Debug.Assert (_patternAnalysisAttempted, "To correctly report all patterns, when starting to analyze a pattern the AnalyzingPattern must be called first.");
```

Comment:
Why not make this a throw?  If all code paths correctly call `AnalyzingPattern` it would never get hit.

Reply:
Not sure what's the pattern in this repo - basically if this happens it's a bug in the linker - so the exception would be something like "Internal linker problem". I'm used to having asserts in such cases, but if the pattern in this repo is to throw - I'll change it.

Reply:
My philosophy is general if something is wrong throw.  A user may have the code that requires hitting some case, but they will have a release build.  If it's a debug assert we'll never seen it.

If @marek-safar has no objections I'd prefer a throw.